includes/abstracts/abstract-wc-shipping-method.php

```diff
@@ -230,7 +230,9 @@ public function get_title() {
 	public function get_rates_for_package( $package ) {
 		$this->rates = array();
 		if ( $this->is_available( $package ) && ( empty( $package['ship_via'] ) || in_array( $this->id, $package['ship_via'] ) ) ) {
+			do_action( 'woocommerce_before_calculate_shipping', $package, $this );
 			$this->calculate_shipping( $package );
+			do_action( 'woocommerce_after_calculate_shipping', $package, $this );
```

Comment:
There's already some hooks that could be used instead of including new ones.

For example each shipping method have a `calculate_shipping` method and will trigger the hook `woocommerce_<shipping_method_id>_shipping_add_rate`, we could introduce a new hook after that more generic as well.

And there's also the `woocommerce_package_rates` filter that could be used to change values or anything you like in the packages.